File: hive/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iceberg.hive;

import com.google.common.collect.Lists;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hive.metastore.TableType;
import org.apache.hadoop.hive.metastore.api.FieldSchema;
import org.apache.hadoop.hive.metastore.api.LockComponent;
import org.apache.hadoop.hive.metastore.api.LockLevel;
import org.apache.hadoop.hive.metastore.api.LockRequest;
import org.apache.hadoop.hive.metastore.api.LockResponse;
import org.apache.hadoop.hive.metastore.api.LockState;
import org.apache.hadoop.hive.metastore.api.LockType;
import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
import org.apache.hadoop.hive.metastore.api.SerDeInfo;
import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
import org.apache.hadoop.hive.metastore.api.Table;
import org.apache.iceberg.BaseMetastoreTableOperations;
import org.apache.iceberg.Schema;
import org.apache.iceberg.TableMetadata;
import org.apache.iceberg.exceptions.AlreadyExistsException;
import org.apache.iceberg.exceptions.CommitFailedException;
import org.apache.iceberg.exceptions.NoSuchTableException;
import org.apache.iceberg.hadoop.HadoopFileIO;
import org.apache.iceberg.io.FileIO;
import org.apache.thrift.TException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * TODO we should be able to extract some more commonalities to BaseMetastoreTableOperations to
 * avoid code duplication between this class and Metacat Tables.
 */
public class HiveTableOperations extends BaseMetastoreTableOperations {
  private static final Logger LOG = LoggerFactory.getLogger(HiveTableOperations.class);

  private final HiveClientPool metaClients;
  private final String database;
  private final String tableName;
  private final Configuration conf;

  private FileIO fileIO;

  protected HiveTableOperations(Configuration conf, HiveClientPool metaClients, String database, String table) {
    this.conf = conf;
    this.metaClients = metaClients;
    this.database = database;
    this.tableName = table;
  }

  @Override
  public FileIO io() {
    if (fileIO == null) {
      fileIO = new HadoopFileIO(conf);
    }

    return fileIO;
  }

  @Override
  protected void doRefresh() {
    String metadataLocation = null;
    try {
      final Table table = metaClients.run(client -> client.getTable(database, tableName));
      String tableType = table.getParameters().get(TABLE_TYPE_PROP);

      if (tableType == null || !tableType.equalsIgnoreCase(ICEBERG_TABLE_TYPE_VALUE)) {
        throw new IllegalArgumentException(String.format("Type of %s.%s is %s, not %s",
            database, tableName,
            tableType /* actual type */, ICEBERG_TABLE_TYPE_VALUE /* expected type */));
      }

      metadataLocation = table.getParameters().get(METADATA_LOCATION_PROP);
      if (metadataLocation == null) {
        String errMsg = String.format("%s.%s is missing %s property", database, tableName, METADATA_LOCATION_PROP);
        throw new IllegalArgumentException(errMsg);
      }

    } catch (NoSuchObjectException e) {
      if (currentMetadataLocation() != null) {
        throw new NoSuchTableException(String.format("No such table: %s.%s", database, tableName));
      }

    } catch (TException e) {
      String errMsg = String.format("Failed to get table info from metastore %s.%s", database, tableName);
      throw new RuntimeException(errMsg, e);

    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new RuntimeException("Interrupted during refresh", e);
    }

    refreshFromMetadataLocation(metadataLocation);
  }

  @Override
  protected void doCommit(TableMetadata base, TableMetadata metadata) {
    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);

    boolean threw = true;
    Optional<Long> lockId = Optional.empty();
    try {
      lockId = Optional.of(acquireLock());
      // TODO add lock heart beating for cases where default lock timeout is too low.
      Table tbl;
      if (base != null) {
        tbl = metaClients.run(client -> client.getTable(database, tableName));
      } else {
        final long currentTimeMillis = System.currentTimeMillis();
        tbl = new Table(tableName,
            database,
            System.getProperty("user.name"),
            (int) currentTimeMillis / 1000,
            (int) currentTimeMillis / 1000,
            Integer.MAX_VALUE,
            storageDescriptor(metadata),
            Collections.emptyList(),
            new HashMap<>(),
            null,
            null,
            TableType.EXTERNAL_TABLE.toString());
        tbl.getParameters().put("EXTERNAL", "TRUE"); // using the external table type also requires this
      }

      tbl.setSd(storageDescriptor(metadata)); // set to pickup any schema changes
      String metadataLocation = tbl.getParameters().get(METADATA_LOCATION_PROP);
      String baseMetadataLocation = base != null ? base.file().location() : null;
      if (!Objects.equals(baseMetadataLocation, metadataLocation)) {
        throw new CommitFailedException(
            "Base metadata location '%s' is not same as the current table metadata location '%s' for %s.%s",
            baseMetadataLocation, metadataLocation, database, tableName);
      }

      setParameters(newMetadataLocation, tbl);

      if (base != null) {
        metaClients.run(client -> {
          client.alter_table(database, tableName, tbl);
          return null;
        });
      } else {
        metaClients.run(client -> {
          client.createTable(tbl);
          return null;
        });
      }
      threw = false;
    } catch (org.apache.hadoop.hive.metastore.api.AlreadyExistsException e) {
      throw new AlreadyExistsException("Table already exists: %s.%s", database, tableName);

    } catch (TException | UnknownHostException e) {
      if (e.getMessage().contains("Table/View 'HIVE_LOCKS' does not exist")) {
        throw new RuntimeException("Failed to acquire locks from metastore because 'HIVE_LOCKS' doesn't " +
            "exist, this probably happened when using embedded metastore or doesn't create a " +
            "transactional meta table. To fix this, use an alternative metastore", e);
      }

      throw new RuntimeException(String.format("Metastore operation failed for %s.%s", database, tableName), e);

    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new RuntimeException("Interrupted during commit", e);

    } finally {
      if (threw) {
        // if anything went wrong, clean up the uncommitted metadata file
        io().deleteFile(newMetadataLocation);
      }
      unlock(lockId);
    }
  }

  private void setParameters(String newMetadataLocation, Table tbl) {
    Map<String, String> parameters = tbl.getParameters();

    if (parameters == null) {
      parameters = new HashMap<>();
    }

    parameters.put(TABLE_TYPE_PROP, ICEBERG_TABLE_TYPE_VALUE.toUpperCase(Locale.ENGLISH));
    parameters.put(METADATA_LOCATION_PROP, newMetadataLocation);

    if (currentMetadataLocation() != null && !currentMetadataLocation().isEmpty()) {
      parameters.put(PREVIOUS_METADATA_LOCATION_PROP, currentMetadataLocation());
    }

    tbl.setParameters(parameters);
  }

  private StorageDescriptor storageDescriptor(TableMetadata metadata) {

    final StorageDescriptor storageDescriptor = new StorageDescriptor();
    storageDescriptor.setCols(columns(metadata.schema()));
    storageDescriptor.setLocation(metadata.location());
    storageDescriptor.setOutputFormat("org.apache.hadoop.mapred.FileOutputFormat");
    storageDescriptor.setInputFormat("org.apache.hadoop.mapred.FileInputFormat");
    SerDeInfo serDeInfo = new SerDeInfo();
    serDeInfo.setSerializationLib("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe");
    storageDescriptor.setSerdeInfo(serDeInfo);
    return storageDescriptor;
  }

  private List<FieldSchema> columns(Schema schema) {
    return schema.columns().stream()
        .map(col -> new FieldSchema(col.name(), HiveTypeConverter.convert(col.type()), ""))
        .collect(Collectors.toList());
  }

  private long acquireLock() throws UnknownHostException, TException, InterruptedException {
    final LockComponent lockComponent = new LockComponent(LockType.EXCLUSIVE, LockLevel.TABLE, database);
    lockComponent.setTablename(tableName);
    final LockRequest lockRequest = new LockRequest(Lists.newArrayList(lockComponent),
        System.getProperty("user.name"),
        InetAddress.getLocalHost().getHostName());
    LockResponse lockResponse = metaClients.run(client -> client.lock(lockRequest));
    LockState state = lockResponse.getState();
    long lockId = lockResponse.getLockid();
    //TODO add timeout
    while (state.equals(LockState.WAITING)) {
      lockResponse = metaClients.run(client -> client.checkLock(lockId));
      state = lockResponse.getState();
      Thread.sleep(50);
    }

    if (!state.equals(LockState.ACQUIRED)) {
      throw new CommitFailedException(String.format("Could not acquire the lock on %s.%s, " +
          "lock request ended in state %s", database, tableName, state));
    }
    return lockId;
  }

  private void unlock(Optional<Long> lockId) {
    if (lockId.isPresent()) {
      try {
        metaClients.run(client -> {
          client.unlock(lockId.get());
          return null;
        });
      } catch (Exception e) {
        throw new RuntimeException(String.format("Failed to unlock %s.%s", database, tableName), e);
      }
    }
  }
}

File: python/iceberg/hive/hive_table_operations.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
import getpass
import logging
import socket
import time
from typing import List

from hmsclient import HMSClient
from hmsclient.genthrift.hive_metastore.ttypes import AlreadyExistsException, EnvironmentContext, FieldSchema, \
    LockComponent, LockLevel, LockRequest, LockResponse, LockState, LockType, NoSuchObjectException, \
    SerDeInfo, StorageDescriptor, Table, TException

from .hive_types import hive_types
from ..api import Schema
from ..api.types import Type
from ..core import BaseMetastoreTableOperations, TableMetadata
from ..core.filesystem import FileSystem, get_fs
from ..exceptions import AlreadyExistsException as IcebergAlreadyExistsException
from ..exceptions import CommitFailedException


class HiveTableOperations(BaseMetastoreTableOperations):

    def __init__(self: "HiveTableOperations", conf: dict, client: HMSClient, database: str, table: str) -> None:
        super(HiveTableOperations, self).__init__(conf)
        self._client = client
        self.database = database
        self.table = table
        try:
            self.refresh()
        except NoSuchObjectException:
            pass  # Object hasn't been created yet, can't refresh.

    def refresh(self: "HiveTableOperations") -> TableMetadata:
        with self._client as open_client:
            tbl_info = open_client.get_table(self.database, self.table)

        table_type = tbl_info.parameters.get(BaseMetastoreTableOperations.TABLE_TYPE_PROP)

        if not table_type or table_type.lower() != BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE:
            raise RuntimeError("Invalid table, not Iceberg: %s.%s" % (self.database,
                                                                      self.table))

        metadata_location = tbl_info.parameters.get(BaseMetastoreTableOperations.METADATA_LOCATION_PROP)
        if not metadata_location:
            raise RuntimeError("Invalid table, missing metadata_location: %s.%s" % (self.database,
                                                                                    self.table))

        self.refresh_from_metadata_location(metadata_location)

        return self.current()

    def commit(self: "HiveTableOperations", base: TableMetadata, metadata: TableMetadata) -> None:
        new_metadata_location = self.write_new_metadata(metadata, self.version + 1)

        threw = True
        lock_id = None
        try:
            lock_id = self.acquire_lock()
            if base:
                with self._client as open_client:
                    tbl = open_client.get_table(self.database, self.table)
            else:
                current_time_millis = int(time.time())
                tbl = Table(self.table,
                            self.database,
                            getpass.getuser(),
                            current_time_millis // 1000,
                            current_time_millis // 1000,
                            sd=storage_descriptor(metadata),
                            tableType="EXTERNAL_TABLE",
                            parameters={'EXTERNAL': 'TRUE'})

            tbl.sd = storage_descriptor(metadata)
            metadata_location = tbl.parameters.get(BaseMetastoreTableOperations.METADATA_LOCATION_PROP, None)
            base_metadata_location = base.metadata_location if base else None
            if base_metadata_location != metadata_location:
                raise CommitFailedException(
                    "Base metadata location '%s' is not same as the current table metadata location '%s' for %s.%s",
                    base_metadata_location, metadata_location, self.database, self.table)

            self.set_parameters(new_metadata_location, tbl)

            if base:
                with self._client as open_client:
                    env_context = EnvironmentContext(
                        {"DO_NOT_UPDATE_STATS": "true"}
                    )
                    open_client.alter_table_with_environment_context(self.database, self.table, tbl, env_context)

            else:
                with self._client as open_client:
                    open_client.create_table(tbl)
            threw = False
        except AlreadyExistsException:
            raise IcebergAlreadyExistsException("Table already exists: {}.{}".format(self.database, self.table))
        except TException as e:
            if e and "Table/View 'HIVE_LOCKS' does not exist" in str(e):
                raise Exception("""Failed to acquire locks from metastore because 'HIVE_LOCKS' doesn't
                                exist, this probably happened when using embedded metastore or doesn't create a
                                transactional meta table. To fix this, use an alternative metastore""", e)

            raise Exception("Metastore operation failed for {}.{}".format(self.database, self.table), e)
        finally:
            if threw:
                self.io().delete(new_metadata_location)
            self.unlock(lock_id)

    def set_parameters(self: "HiveTableOperations", new_metadata_location: str, tbl: Table) -> None:
        parameters = tbl.parameters

        if not parameters:
            parameters = dict()

        parameters[BaseMetastoreTableOperations.TABLE_TYPE_PROP] = \
            BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE.upper()
        parameters[BaseMetastoreTableOperations.METADATA_LOCATION_PROP] = new_metadata_location

        if self.current_metadata_location and len(self.current_metadata_location) > 0:
            parameters[BaseMetastoreTableOperations.PREVIOUS_METADATA_LOCATION_PROP] = self.current_metadata_location

        tbl.parameters = parameters

    def unlock(self: "HiveTableOperations", lock_id: int = None) -> None:
        if lock_id:
            try:
                with self._client as open_client:
                    open_client.unlock(LockResponse(lock_id))
            except Exception as e:
                logging.warning("Failed to unlock {}.{}".format(self.database, self.table), e)

    def acquire_lock(self: "HiveTableOperations") -> int:
        lock_component = LockComponent(LockType.EXCLUSIVE, LockLevel.TABLE, self.database, self.table)

        lock_request = LockRequest([lock_component], user=getpass.getuser(), hostname=socket.gethostname())
        with self._client as open_client:
            lock_response = open_client.lock(lock_request)

        state = lock_response.state
        lock_id = lock_response.lockid
        start = int(time.time())
        duration = 0
        timed_out = False
        while not timed_out and state == LockState.WAITING:
            with self._client as open_client:
                lock_response = open_client.check_lock(lock_response)
            state = lock_response.state

            duration = int(time.time()) - start
            if duration > 3 * 60 * 1000:
                timed_out = True
            else:
                time.sleep(0.05)

        if timed_out and state != LockState.ACQUIRED:
            raise CommitFailedException("Timed out after {} ms waiting for lock on {}.{}".format(duration,
                                                                                                 self.database,
                                                                                                 self.table))

        if state != LockState.ACQUIRED:
            raise CommitFailedException(
                "Could not acquire the lock on {}.{}, lock request ended in state {}".format(self.database, self.table,
                                                                                             state))
        return lock_id

    def io(self: "HiveTableOperations") -> FileSystem:
        return get_fs(self.base_location, self.conf)

    def close(self: "HiveTableOperations") -> None:
        self._client.close()


def storage_descriptor(metadata: TableMetadata) -> StorageDescriptor:
    ser_de_info = SerDeInfo(serializationLib="org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe")
    return StorageDescriptor(columns(metadata.schema),
                             metadata.location,
                             "org.apache.hadoop.mapred.FileInputFormat",
                             "org.apache.hadoop.mapred.FileOutputFormat",
                             serdeInfo=ser_de_info)


def columns(schema: Schema) -> List[FieldSchema]:
    return [FieldSchema(col.name, convert_hive_type(col.type), "") for col in schema.columns()]


def convert_hive_type(col_type: Type) -> str:
    type_id = col_type.type_id
    hive_type_id = hive_types.get(type_id)  # type: ignore
    if hive_type_id:
        return hive_type_id
    raise NotImplementedError("Not yet implemented column type " + str(col_type))
